framework/src/hdgkernels/DiffusionHDGAssemblyHelper.C

```diff
@@ -0,0 +1,403 @@
+                                                       MaterialPropertyInterface * const mpi,
+                                                       const TransientInterface * const ti,
+                                                       SystemBase & nl_sys,
+                                                       SystemBase & aux_sys,
+                                                       const THREAD_ID tid)
+  : // vars
+    _u_var(aux_sys.getFieldVariable<Real>(tid, moose_obj->getParam<AuxVariableName>("u"))),
+    _grad_u_var(aux_sys.getFieldVariable<RealVectorValue>(
+        tid, moose_obj->getParam<AuxVariableName>("grad_u"))),
+    _u_face_var(
+        nl_sys.getFieldVariable<Real>(tid, moose_obj->getParam<NonlinearVariableName>("face_u"))),
+    // dof indices
+    _qu_dof_indices(_grad_u_var.dofIndices()),
+    _u_dof_indices(_u_var.dofIndices()),
+    _lm_u_dof_indices(_u_face_var.dofIndices()),
+    // solutions
+    _qu_sol(_grad_u_var.sln()),
+    _u_sol(_u_var.sln()),
+    _lm_u_sol(_u_face_var.sln()),
+    // shape functions
+    _vector_phi(_grad_u_var.phi()),
+    _scalar_phi(_u_var.phi()),
+    _grad_scalar_phi(_u_var.gradPhi()),
+    _div_vector_phi(_grad_u_var.divPhi()),
+    _vector_phi_face(_grad_u_var.phiFace()),
+    _scalar_phi_face(_u_var.phiFace()),
+    _lm_phi_face(_u_face_var.phiFace()),
+    // material properties
+    _diff(mpi->getMaterialProperty<Real>("diffusivity")),
+    // initialize local number of dofs
+    _vector_n_dofs(0),
+    _scalar_n_dofs(0),
+    _lm_n_dofs(0),
+    _ti(*ti),
+    _tau(moose_obj->getParam<Real>("tau"))
+{
+}
+
+std::set<const MooseVariableBase *>
+DiffusionHDGAssemblyHelper::variables() const
+{
+  return {&_u_var, &_grad_u_var, &_u_face_var};
+}
+
+void
+DiffusionHDGAssemblyHelper::resizeData()
+{
+  _vector_n_dofs = _qu_dof_indices.size();
+  _scalar_n_dofs = _u_dof_indices.size();
+  _lm_n_dofs = _lm_u_dof_indices.size();
+
+  libmesh_assert_equal_to(_vector_n_dofs, _vector_phi.size());
+  libmesh_assert_equal_to(_scalar_n_dofs, _scalar_phi.size());
+
+  _primal_size = _vector_n_dofs + _scalar_n_dofs;
+  _lm_size = _lm_n_dofs;
+
+  // prepare our matrix/vector data structures
+  _PrimalMat.setZero(_primal_size, _primal_size);
+  _PrimalVec.setZero(_primal_size);
+  _LMMat.setZero(_lm_size, _lm_size);
+  _LMVec.setZero(_lm_size);
+  _PrimalLM.setZero(_primal_size, _lm_size);
+  _LMPrimal.setZero(_lm_size, _primal_size);
+}
+
+void
+DiffusionHDGAssemblyHelper::vectorVolumeResidual(const unsigned int i_offset,
+                                                 const MooseArray<Gradient> & vector_sol,
+                                                 const MooseArray<Number> & scalar_sol,
+                                                 const MooseArray<Real> & JxW,
+                                                 const QBase & qrule)
+{
+  for (const auto qp : make_range(qrule.n_points()))
+    for (const auto i : make_range(_vector_n_dofs))
```

Comment:
I would guess flipping these two loops would be faster. seems like it would be vectorizable over Qps even

but only a guess